docs/contract/confidential-smart-contract.md

```diff
@@ -0,0 +1,264 @@
+
+    fn query<C: sdk::Context>(_ctx: &mut C, _request: Request) -> Result<Response, Error> {
+        // This method is called for each contracts.Query query. It is supposed to handle the
+        // request and return a response.
+        Err(Error::BadRequest)
+    }
+}
+
+// Create the required WASM exports required for the contract to be runnable.
+sdk::create_contract!(HelloWorld);
+
+// We define some simple contract tests below.
+#[cfg(test)]
+mod test {
+    use oasis_contract_sdk::{testing::MockContext, types::ExecutionContext, Contract};
+
+    use super::*;
+
+    #[test]
+    fn test_hello() {
+        // Create a mock execution context with default values.
+        let mut ctx: MockContext = ExecutionContext::default().into();
+
+        // Instantiate the contract.
+        HelloWorld::instantiate(
+            &mut ctx,
+            Request::Instantiate {
+                initial_counter: 11,
+            },
+        )
+        .expect("instantiation should work");
+
+        // Dispatch the SayHello message.
+        let rsp = HelloWorld::call(
+            &mut ctx,
+            Request::SayHello {
+                who: "unit test".to_string(),
+            },
+        )
+        .expect("SayHello call should work");
+
+        // Make sure the greeting is correct.
+        assert_eq!(
+            rsp,
+            Response::Hello {
+                greeting: "hello unit test (11)".to_string()
+            }
+        );
+
+        // Dispatch another SayHello message.
+        let rsp = HelloWorld::call(
+            &mut ctx,
+            Request::SayHello {
+                who: "second call".to_string(),
+            },
+        )
+        .expect("SayHello call should work");
+
+        // Make sure the greeting is correct.
+        assert_eq!(
+            rsp,
+            Response::Hello {
+                greeting: "hello second call (12)".to_string()
+            }
+        );
+    }
+}
+```
+
+The contract is built the same way as its non-confidential counterpart:
+
+```shell
+cargo build --target wasm32-unknown-unknown --release
+```
+
+::: warning
+
+The blockchain store containing all compiled contracts is public. This means
+that anyone will be able to decompile your smart contract and see how it
+works. **Do not put any sensitive data inside the smart contract code!** 
+
+:::
+
+Since the smart contracts store is public, uploading the WASM code is
+the same as for the non-confidential ones:
+
+```shell
+oasis contracts upload hello_world.wasm
+```
+
+## Confidential Instantiation and Calling
+
+To generate a confidential transaction, the `oasis contracts` subcommand
+accepts `--encrypted` flag. Confidential transactions have encrypted contract
+address, function name, parameters and the amounts and types of tokens sent.
+**However, the _authorization information_ which contains information on the
+signer is public!** Usually this field contains the public key of your account
+or a list of expected multisig keys.
```

Comment:
Together with the gas limit and amount of fee to be paid for processing the transaction.